Mathlib/Data/Seq/Computation.lean

```diff
@@ -0,0 +1,1296 @@
+def destruct (c : Computation α) : Sum α (Computation α) :=
+  match c.1 0 with
+  | none => Sum.inr (tail c)
+  | some a => Sum.inl a
+#align computation.destruct Computation.destruct
+
+/-- `run c` is an unsound meta function that runs `c` to completion, possibly
+  resulting in an infinite loop in the VM. -/
+unsafe def run : Computation α → α
+  | c =>
+    match destruct c with
+    | Sum.inl a => a
+    | Sum.inr ca => run ca
+#align computation.run Computation.run
+
+theorem destruct_eq_pure {s : Computation α} {a : α} : destruct s = Sum.inl a → s = pure a := by
+  dsimp [destruct]
+  induction' f0 : s.1 0 with _ <;> intro h
+  · contradiction
+  · apply Subtype.eq
+    funext n
+    induction' n with n IH
+    · injection h with h'
+      rwa [h'] at f0
+    · exact s.2 IH
+#align computation.destruct_eq_ret Computation.destruct_eq_pure
+
+theorem destruct_eq_think {s : Computation α} {s'} : destruct s = Sum.inr s' → s = think s' := by
+  dsimp [destruct]
+  induction' f0 : s.1 0 with a' <;> intro h
+  · injection h with h'
+    rw [← h']
+    cases' s with f al
+    apply Subtype.eq
+    dsimp [think, tail]
+    rw [← f0]
+    exact (Stream'.eta f).symm
+  · contradiction
+#align computation.destruct_eq_think Computation.destruct_eq_think
+
+@[simp]
+theorem destruct_pure (a : α) : destruct (pure a) = Sum.inl a :=
+  rfl
+#align computation.destruct_ret Computation.destruct_pure
+
+@[simp]
+theorem destruct_think : ∀ s : Computation α, destruct (think s) = Sum.inr s
+  | ⟨_, _⟩ => rfl
+#align computation.destruct_think Computation.destruct_think
+
+@[simp]
+theorem destruct_empty : destruct (empty α) = Sum.inr (empty α) :=
+  rfl
+#align computation.destruct_empty Computation.destruct_empty
+
+@[simp]
+theorem head_pure (a : α) : head (pure a) = some a :=
+  rfl
+#align computation.head_ret Computation.head_pure
+
+@[simp]
+theorem head_think (s : Computation α) : head (think s) = none :=
+  rfl
+#align computation.head_think Computation.head_think
+
+@[simp]
+theorem head_empty : head (empty α) = none :=
+  rfl
+#align computation.head_empty Computation.head_empty
+
+@[simp]
+theorem tail_pure (a : α) : tail (pure a) = pure a :=
+  rfl
+#align computation.tail_ret Computation.tail_pure
+
+@[simp]
+theorem tail_think (s : Computation α) : tail (think s) = s := by
+  cases' s with f al ; apply Subtype.eq ; dsimp [tail, think] ; rw [Stream'.tail_cons]
+#align computation.tail_think Computation.tail_think
+
+@[simp]
+theorem tail_empty : tail (empty α) = empty α :=
+  rfl
+#align computation.tail_empty Computation.tail_empty
+
+theorem think_empty : empty α = think (empty α) :=
+  destruct_eq_think destruct_empty
+#align computation.think_empty Computation.think_empty
+
+/-- Recursion principle for computations, compare with `List.recOn`. -/
+-- porting notes: required `noncomputable`
+noncomputable def recOn {C : Computation α → Sort v} (s : Computation α) (h1 : ∀ a, C (pure a))
```

Comment:
Can we make this computable again, using the approach of `List.recOn'` or something similar?